### ISSUE TITLE
`isinstance` with generic type

### ISSUE THREAD

**Issue by jhance** [python/mypy#949]
For example

```
class A(Generic[T]):
   pass

class B(Generic[T], A[T]):
   pass

def f(x: A[T]):
    if isinstance(x, B[T]):
        ...
```

is apparently illegal and I'm not sure what replacement I'm supposed to use. It tells me to use a `# type` comment which isn't really applicable here.

Removing the `[T]` causes mypy to be unable to infer that `x` is in fact a `B[T]`.

**Comment by JukkaL** [comment]
Hmm this looks like a mypy bug. It should work like this:

```
class A(Generic[T]):
   pass

class B(Generic[T], A[T]):
   pass

def f(x: A[T]):
    if isinstance(x, B):
        # type of B is B[T] here
        ...
```

**Comment by JukkaL** [comment]
And if you try to use `isinstance(x, B[T])`, mypy should give a useful error message.

**Comment by jhance** [comment]
I don't think that is powerful enough.

What if I have an `A[U,V]` and want to know if its an `A[U,U]`?

**Comment by JukkaL** [comment]
We are stuck with `isinstance` semantics here. However, `isinstance` doesn't seem to be used that often with generic types, at least in the code I've looked at.

**Comment by jhance** [comment]
The issue is there logically isn't anything we can infer here, as far as I can tell. Since `isinstance` does not respect generics, we can't either. The best we can do is give an error message instructing the user that they need to cast.

**Comment by elazarg** [comment]
Related #3037

**Comment by ilevkivskyi** [comment]
On current master it behaves like this:
```python
class A(Generic[T]):
   pass

class B(A[T]):
   pass

x: A[int]
if isinstance(x, B):
    reveal_type(x)  # Revealed type is '__main__.B[Any]'
isinstance(x, B[int])  # Parameterized generics cannot be used with class or instance checks
```
It looks like the issue is fixed only partially. I think the revealed type should be ``B[int]`` (even if it is not 100% safe).

**Comment by sobolevn** [comment]
Are there any known work-arounds?

**Comment by Stewori** [comment]
For the record, typechecking of generics *does* work in pytypes (Python 3.7+ support is almost done but not released as of this writing). The issue linked above rather required type inference which is beyond the scope of pytypes. See the concrete [example](https://github.com/Stewori/pytypes/issues/82#issuecomment-546603223) that was requested and the answers for clarification.

**Comment by AlexWaygood** [comment]
In 2022, the way to achieve "slightly unsafe" type-narrowing of generics in this fashion is to use [`typing.TypeGuard`](https://docs.python.org/3/library/typing.html#typing.TypeGuard) (or `typing_extensions.TypeGuard` on Python <=3.9).
